File: Java/9.palindrome-numbers.java
class Solution {
    public boolean isPalindrome(int x) {
          // Negative numbers and numbers ending with 0 (except 0 itself) are not palindromic
        if (x < 0 || (x % 10 == 0 && x != 0)) {
            return false;
        }

        int reversedNumber = 0;

        // Reverse only the second half of the number
        while (x > reversedNumber) {
            int digit = x % 10;
            reversedNumber = reversedNumber * 10 + digit;
            x /= 10;
        }

        // For odd-length numbers, remove the middle digit from the reversed number
        return x == reversedNumber || x == reversedNumber / 10;
    }
}

File: Rust/9.palindrome-numbers.rs
impl Solution {
    pub fn is_palindrome(x: i32) -> bool {
        // Negative numbers are not palindromic
        if x < 0 {
            return false;
        }

        let original_number = x;
        let mut reversed_number = 0;

        let mut x_mut = x;
        while x_mut > 0 {
            // Extract the last digit of the number
            let digit = x_mut % 10;

            // Reverse the number by adding one digit at a time
            reversed_number = reversed_number * 10 + digit;

            // Remove the rightmost digit from the number
            x_mut /= 10;
        }

        // Check if the reversed number is equal to the original number
        original_number == reversed_number
    }
}

fn main() {
    // Example usage
    println!("{}", Solution::is_palindrome(121));  // true
    println!("{}", Solution::is_palindrome(-121)); // false
    println!("{}", Solution::is_palindrome(10));   // false
}